Sources/GithubAPI.swift

```diff
@@ -21,74 +21,111 @@ import PerfectLogger
+    var value: GithubAPI?
+    githubInstancesLock.doWithReadLock {
+      value = githubInstances[installation]
+    }
+    return value
+  }
+
+  func getGithubInstance(for installation: String) -> GithubAPI? {
+    if let githubInstanceCached = getCachedGithubInstance(for: installation) {
+      return githubInstanceCached
+    }
+
+    return githubInstancesLock.doWithWriteLock { () -> GithubAPI? in
+      guard let accessToken = GithubAuth.getAccessToken(installationID: installation) else {
+        LogFile.error("couldn't get an access token for installation: \(installation)")
+        return nil
+      }
+      let githubInstance = GithubAPI(accessToken: accessToken, installationID: installation)
+      githubInstances[accessToken] = githubInstance
+      return githubInstance
+    }
+  }
+}
 
+class GithubCURLRequest: CURLRequest {
   override init(options: [CURLRequest.Option]) {
     super.init(options: options)
     var action = "unknown action"
     outer: for option in options {
       switch option {
-        case .url(let urlString):
+      case .url(let urlString):
         action = urlString
         break outer
       default:
         break
       }
     }
     Analytics.trackEvent(category: "Github API", action: action)
-    GithubCURLRequest.curlAccessLock.lock()
-    if time(nil) - GithubAPI.lastGithubAccess < 1 {
-      Threading.sleep(seconds: 1)
-    }
-    GithubAPI.lastGithubAccess = time(nil)
-    GithubCURLRequest.curlAccessLock.unlock()
   }
-
 }
 
 public class GithubAPI {
-  static let githubBaseURL = "https://api.github.com"
-  static let retryCount = 0
-  static var lastGithubAccess = time(nil)
 
+  var accessToken: String
+  var installationID: String
+  let curlAccessLock = Threading.Lock()
+  var lastGithubAccess = time(nil)
+
+  init(accessToken: String, installationID: String) {
+    self.accessToken = accessToken
+    self.installationID = installationID
+  }
 
   /// This method adds labels to a Github issue through the API.
   ///
   /// - Parameters:
   ///   - url: The url of the issue as a String
   ///   - labels: The labels to add to the issue
-  class func addLabelsToIssue(url: String, labels: [String]) {
+  func addLabelsToIssue(url: String, labels: [String]) {
     LogFile.debug(labels.description)
-    let labelsURL = url + "/labels"
-    do {
+
+    let performRequest = { () -> CURLResponse in
+      let labelsURL = url + "/labels"
       let request = GithubCURLRequest(labelsURL, .postString(labels.description))
-      addAPIHeaders(to: request)
-      let response = try request.perform()
-      if GithubAuth.refreshCredentialsIfUnauthorized(response: response) {
-        addLabelsToIssue(url: url, labels: labels)
+      self.addAPIHeaders(to: request)
+      return try request.perform()
+    }
+
+    do {
+      var response = try performRequest()
+      if GithubAuth.refreshCredentialsIfUnauthorized(response: response, githubInstance: self) {
+        response = try performRequest()
       }
```

Comment:
Can we capture this pattern in a shared location?

Reply:
done!